### DISCUSSION TITLE
Loading HEX files into EEPROM & LUTs

### DISCUSSION THREAD

**Discussion by ddotcole** [hneemann/Digital#901] [Ideas]
Hello,

I have a simulation going with 8 different EEPROMS & LUTs.  Its a bit tedious manually loading each HEX file anytime I update the input HEX files.  Is there a way where the files could be linked to the device and anytime you start a simulation it updates the device contents automatically from the HEX files on my harddrive?

Hopefully I have explained this well enough.

**Comment by richardson1959** [comment]
This is probably not the answer you are looking for, but the .dig files are xml files.  The ROM components appear to have their contents included as run-length-encoded text.  You could create embedded modules (wrappers) around each ROM and then write a program/script to update the corresponding dig files when ever you need to.  (I'd recommend using embedded modules just so that you aren't updating the one big dig file containing all of your ROMs.)  I agree that it would be nicer to have the option to have the contents linked to a binary file containing the data.

**Comment by ddotcole** [comment]
I have been working on a script with the ability to take my hex files, turn them into the run length encoding used in Digital and replace the data for each LUT and ROM in the .dig file using a tool called xmlstartlet.  It is pretty slick to run through .dig file with xmlstartlet and labeling the LUTs and ROMs is key.  This allows you to find the right spot to replace the data, which should be trivial to produce as well.

**Reply by ddotcole** [reply]
If there are no bytes with data in either the ROMs or LUTs, the ROM and LUTs don't show up in the .dig file.  This is a more difficult process then I expected.

**Reply by hneemann** [reply]
Default values are not stored in the dig file to keep the file small. And if all values in a LUT or a memory are 0, this is the default value and therefore is missing in the file.
But can't you just enter a 1, and then replace it?

**Reply by ddotcole** [reply]
Yes, that is what I ended up doing.  My script checks if the run length encoding ends up all zeros it throws a 1 in at the end so as to never be empty.  This has kept them in the file so my script works.  Less repetitive stress now.

**Comment by richardson1959** [comment]
The ROM component has the option to load it's contents from a file.  It is a raw binary file.  Can you use ROMs instead of EEPROMs?  If not, you could have an initialization circuit that initializes each EEPROM from a ROM ?  You can't initialize the lookup table components this way, but you could use a ROM as a LUT.

**Reply by ddotcole** [reply]
Interesting thought.  I have my script replacement working well now.  I just close the program, run my script to update the .dig file then open it back up in Digital.  The program opens pretty fast as well so its a quick enough process now with much less mouse clicking.
